### PR TITLE
func: add basically type cast overflow test #2

### PR DESCRIPTION
**Pull request by zintown**

### CODE REVIEW

**Comment by wsong83** [comment]
First of all, remove all NDBUG related code please. DONT do it.
Although it might against your understanding, maintaing the code is already tough enough.
No body wants to maintain the debug code as well.

**Comment by zintown** [comment]
Ok, I will remove the NDBUG part of the code instantly

File: configure.json [hunk 1]
```diff
@@ -1,4 +1,24 @@
 {
+    "mss-overflow-after-typecast-for-basictype-read-ptr-stack": {
+        "program": "mss-read-after-typecast-for-basictype",
+        "arguments": { "0": ["0"] },
+        "expect-results": { "1": "defend enabled."}
+    },
```
Comment (zintown) [inline configure.json:6]: 11111111111111

File: mss/read-after-typecast-for-basictype.cpp [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+      tmp = (*ptr) >> 8*sizeof(char);
+    }else{
+      tmp = *ptr;
+    }
+
+    int devi = 8*(sizeof(char)-1);
+
+    for(size_t i= 0; i < len; i+=step){
+      char cmp = static_cast<char>(tmp & ((0xff << devi) & 0xffffffff));
+      if(cmp != c) return 1;
```
[the remark this answers is not in the document]
Comment (zintown) [reply mss/read-after-typecast-for-basictype.cpp:63]: I am not sure, My design is to test whether variable inside a object/structure is readable, this program trys to type-cast the last char in buffer_data.data  into a long long variable(means 1 byte in data and other 7 bytes in overflow). And I use bit operation to get  each byte in long long, then compare it with the char in buffer_data.overflow.Finally, we can judge whether the pointer can read the information.

File: mss/read-after-typecast-for-basictype.cpp [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+  case 2:{
+    return read_after_typecast(std::end(buffer_data.data)-char_len,char_len,ll_len-char_len,char_len,'d');
+  }
+  case 3:{
+    return read_after_typecast(std::end(buffer_rodata.data)-char_len,char_len,ll_len-char_len,char_len,'d');
+  }
+  }
+}
+
+static int read_after_typecast(const char* cb, size_t offset, size_t len, size_t step, char c){
```
[the remark this answers is not in the document]
Comment (zintown) [reply mss/read-after-typecast-for-basictype.cpp:47]: OK,I revised the above points you mentioned in the latest submission

File: mss/read-after-typecast-for-basictype.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (zintown) [reply mss/read-after-typecast-for-basictype.cpp]: > For example, cast a char inside buffer_data.data into a long long which covers space belonging to buffer_data.overflow. In this case, write the long long actually corrupt buffer_data.overflow, crossing the variable boundary!

Yes, I add this in the write-after-typecast-for-basictype test.

**Comment by wsong83** [comment]
@zintown Break down your PR! A single type of test cases in each PR please. DONT put multiple test cases inside a single PR because it normally make the PR more difficult to be merged.

**Comment by zintown** [comment]
> 

Ohh sorry, I forgot this, I will cherry-pick it to another branch.

**Comment by wsong83** [comment]
Totally confused. Please redo it cleanly.

**Pull request closed without merging**